crates/re_chunk/src/batcher.rs

```diff
@@ -0,0 +1,1588 @@
+        fn new(entity_path: EntityPath) -> Self {
+            Self {
+                entity_path,
+                latest: Instant::now(),
+                pending_rows: Default::default(),
+                pending_num_bytes: Default::default(),
+            }
+        }
+
+        fn reset(&mut self) {
+            self.latest = Instant::now();
+            self.pending_rows.clear();
+            self.pending_num_bytes = 0;
+        }
+    }
+
+    let mut accs: IntMap<EntityPath, Accumulator> = IntMap::default();
+
+    // TODO(#4184): do_push_chunk
+
+    fn do_push_row(acc: &mut Accumulator, row: PendingRow) {
+        acc.pending_num_bytes += row.total_size_bytes();
+        acc.pending_rows.push(row);
+    }
+
+    fn do_flush_all(
+        acc: &mut Accumulator,
+        tx_chunk: &Sender<Chunk>,
+        reason: &str,
+        max_chunk_rows_if_unsorted: u64,
+    ) {
+        let rows = std::mem::take(&mut acc.pending_rows);
+        if rows.is_empty() {
+            return;
+        }
+
+        re_log::trace!(
+            "Flushing {} rows and {} bytes. Reason: {reason}",
+            rows.len(),
+            re_format::format_bytes(acc.pending_num_bytes as _)
+        );
+
+        let chunks =
+            PendingRow::many_into_chunks(acc.entity_path.clone(), max_chunk_rows_if_unsorted, rows);
+        for chunk in chunks {
+            let chunk = match chunk {
+                Ok(chunk) => chunk,
+                Err(err) => {
+                    re_log::error!(%err, "corrupt chunk detected, dropping");
+                    continue;
+                }
+            };
+
+            // NOTE: This can only fail if all receivers have been dropped, which simply cannot happen
+            // as long the batching thread is alive… which is where we currently are.
+            tx_chunk.send(chunk).ok();
+        }
+
+        acc.reset();
+    }
+
+    re_log::trace!(
+        "Flushing every: {:.2}s, {} rows, {}",
+        config.flush_tick.as_secs_f64(),
+        config.flush_num_rows,
+        re_format::format_bytes(config.flush_num_bytes as _),
+    );
+
+    use crossbeam::select;
+    loop {
+        select! {
+            recv(rx_cmd) -> cmd => {
+                let Ok(cmd) = cmd else {
+                    // All command senders are gone, which can only happen if the
+                    // `ChunkBatcher` itself has been dropped.
+                    break;
+                };
+
+                match cmd {
+                    Command::AppendRow(entity_path, row) => {
+                        let acc = accs.entry(entity_path.clone())
+                            .or_insert_with(|| Accumulator::new(entity_path));
+                        do_push_row(acc, row);
+
+                        if let Some(config) = config.hooks.on_insert.as_ref() {
+                            config(&acc.pending_rows);
+                        }
+
+                        if acc.pending_rows.len() as u64 >= config.flush_num_rows {
```

Comment:
A nice side-effect of this refactor is the potential for per-entity-flushing config in the future.